### PR TITLE
Read granted qos for subscribe action and invoke onSubscribeResult

### PR DESCRIPTION
**Pull request by rubenquadros12**
When subscribe action is performed and the result is received read the granted qos array and create success topic list and failed topic list. 
Invoke the new `onSubscribeResult` callback using these 2 lists.

### CODE REVIEW

File: org.eclipse.paho.android.service/build.gradle [hunk 1]
```diff
@@ -42,7 +42,8 @@ configurations {
 
 dependencies {
     implementation "com.android.support:support-v4:${rootProject.ext.supportLibVersion}"
-    implementation "in.mohalla.paho:in.mohalla.paho.client.mqttv3:1.2.7.timer-ping-sender"
+    //implementation "in.mohalla.paho:in.mohalla.paho.client.mqttv3:1.2.7.timer-ping-sender"
```
Comment (rubenquadros12) [inline org.eclipse.paho.android.service/build.gradle]: TODO: revert this change and delete the jar file once https://github.com/ShareChat/paho.mqtt.java/pull/4 is done

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java [hunk 2]
```diff
@@ -1526,6 +1528,28 @@ private void simpleAction(IMqttToken token, Bundle data) {
...
+				for (int i = 0; i < topics.length; i++) {
+					if (grantedQos[i] == 128) {
+						failedTopics.add(topics[i]);
+					} else {
+						successTopics.add(topics[i]);
+					}
+				}
+			}
+
+			((MqttTokenAndroid) token).notifySubscription(successTopics, failedTopics);
```
Comment (dilraj-singh1997) [inline org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java:1548]: should we check `instanceof` here?

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java [hunk 3]
```diff
@@ -1526,6 +1528,28 @@ private void simpleAction(IMqttToken token, Bundle data) {
...
 
+	private void subscribeResultAction(IMqttToken token, Bundle data) {
+		if (token != null) {
+			int[] grantedQos = data.getIntArray(MqttServiceConstants.GRANTED_QOS);
+			String[] topics = token.getTopics();
+			List<String> successTopics = new ArrayList<>();
+			List<String> failedTopics = new ArrayList<>();
+			if (grantedQos != null) {
+				for (int i = 0; i < topics.length; i++) {
+					if (grantedQos[i] == 128) {
```
Comment (pulakdp) [inline org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java:1540]: Do we know for sure that `grantedQos` will be of same length as `token.getTopics()`? // just asking

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java (hunk 2 above)
Comment (rubenquadros12) [reply org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java:1548]: This function is only called from `subscribeAction` which is inside `onReceive` of SUBSCRIBE_ACTION.. hence have not added any additional checks

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java (hunk 3 above)
Comment (rubenquadros12) [reply org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java:1540]: Yes it should be

**Review by rajat-sharechat — APPROVED** [review]

**Review by Swatimahanag — APPROVED** [review]

**Review by pulakdp — APPROVED** [review]

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java (hunk 3 above)
Comment (Abdul-Rizwan) [reply org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java:1540]: @rubenquadros12 please make that `128` as constant variable, then it would be readable.

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttTokenAndroid.java [hunk 4]
```diff
@@ -126,6 +128,17 @@ void notifyComplete() {
     }
   }
 
+  /**
+   * notify successful completion of the subscribe operation
+   */
+  void notifySubscription(List<String> successTopics, List<String> failedTopics) {
+    synchronized (waitObject) {
```
Comment (Abdul-Rizwan) [inline org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttTokenAndroid.java:135]: @rubenquadros12 what `waitObject` does here?

Comment (rubenquadros12) [reply org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttTokenAndroid.java:135]: It is used to lock the thread during operation. This is an existing codebase so have replicated the method.

File: org.eclipse.paho.android.service/build.gradle (hunk 1 above)
Comment (anshul-sharechat) [reply org.eclipse.paho.android.service/build.gradle]: @rubenquadros12 Please revert now

**Review by hiteshchopra11 — APPROVED** [review]

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java [hunk 5]
```diff
@@ -1526,6 +1528,28 @@ private void simpleAction(IMqttToken token, Bundle data) {
...
 	}
 
+	private void subscribeResultAction(IMqttToken token, Bundle data) {
+		if (token != null) {
+			int[] grantedQos = data.getIntArray(MqttServiceConstants.GRANTED_QOS);
+			String[] topics = token.getTopics();
+			List<String> successTopics = new ArrayList<>();
+			List<String> failedTopics = new ArrayList<>();
+			if (grantedQos != null) {
+				for (int i = 0; i < topics.length; i++) {
```
Comment (anshul-sharechat) [inline org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java]: can use foreach here or can make topics.length as constant so that it is not recomputed
https://stackoverflow.com/questions/256859/is-there-a-performance-difference-between-a-for-loop-and-a-for-each-loop

**Review by Abdul-Rizwan — APPROVED** [review]

**Review by dilraj-singh1997 — APPROVED** [review]

File: org.eclipse.paho.android.service/build.gradle (hunk 1 above)
Comment (rubenquadros12) [reply org.eclipse.paho.android.service/build.gradle]: Done

File: org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java (hunk 5 above)
Comment (rubenquadros12) [reply org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/MqttAndroidClient.java]: Have taken `topics.length` in a separate variable. Index is required to get granted qos hence foreach loop is not feasible

**Review by anshul-sharechat — APPROVED** [review]
